UTF-8 uppercase conversion

eshishki opened the pull request:
## Summary

Adds full Unicode UTF-8 uppercase support (`sz_utf8_case_upper`) with serial, AVX-512 (Ice Lake), and AVX2 (Haswell) backends. Along the way also extends and aligns the existing `sz_utf8_case_fold` paths, fixes a few correctness bugs in the existing fold code, and unifies pipeline ordering across all four functions.

Bit-perfect across all 1.1M valid Unicode 16.0 codepoints — `serial ≡ haswell ≡ ice ≡ CPython 3.14 str.upper() / str.casefold()`.

## What's new

### `sz_utf8_case_upper` — new function

Per-codepoint mapper `sz_unicode_upper_codepoint_` covers:
- Simple uppercase from UnicodeData.txt (~1500 one-to-one mappings)
- Unconditional Full mappings from SpecialCasing.txt (ß→SS, ﬃ→FFI, ΐ→Ϊ́, ŉ→ʼN, etc.)
- Conditional mappings (Final_Sigma, Turkic, Lithuanian) are intentionally NOT applied — context-free function

Three backend implementations matching `sz_utf8_case_fold`:
- `_serial` — portable C fallback
- `_ice` — AVX-512 (BW + DQ + VL + VBMI + VBMI2)
- `_haswell` — AVX2 (new, see below)
- `_neon` — stub delegating to `_serial` (matches existing fold pattern)

Public `sz_utf8_case_upper` with dynamic dispatch (`ICE → HASWELL → SERIAL`).

### Haswell (AVX2) backend for `sz_utf8_case_upper` AND `sz_utf8_case_fold`

Previously both functions only had AVX-512 (`_ice`) SIMD; on AMD Zen 1-3 and Intel Haswell..Cascade Lake they silently fell back to serial. Now there's full AVX2 coverage with per-script branches:
- ASCII bulk transform
- Latin-1 (with inline ß↔SS/ss and ÿ↔Ÿ tricks)
- Cyrillic basic (D0/D1)
- Greek (CE/CF) — full bicameral set including tonos and final sigma
- Cyrillic Supplement (D1-ext/D2/D3/D4 combined, upper only)
- 3-byte caseless fast path for CJK/Hangul/Thai/Hindi/etc.
- Georgian Nuskhuri ↔ Asomtavruli
- Georgian Mkhedruli ↔ Mtavruli (+ Mtavruli passthrough)
- Latin Extended Additional (Vietnamese + parity)
- Fullwidth (EF BC ↔ EF BD)

Helpers extracted into `sz_haswell_*` family (`in_range`, `first_invalid`, `byte_mask_shl1/shr1`, `mask_sub/add/set1`, `prefix_bytes`) to keep per-script code readable.

### Extensions to existing `sz_utf8_case_fold_ice`

The upstream fold path didn't cover two scripts I needed for parity with `upper`:
- **Cyrillic Supplement** (D1-ext/D2/D3/D4 fold) — used to fall into the universal codepoint-unpack path
- **Mtavruli → Mkhedruli** (E1 B2 → E1 83) — used to fall into per-rune serial drain

Both now have dedicated SIMD branches in `fold_ice`. Same branches added in `fold_haswell`.

## Bug fixes to existing code

- **`sz_unicode_fold_codepoint_`**: removed 28 bogus mappings for Unicode 16.0 unassigned codepoints (`A7CE→A7CF`, `A7D2→A7D3`, `A7D4→A7D5`, and the entire fake `0x16EA0..0x16EB8` "Beria Erfe" block)
- **`sz_utf8_case_upper_ice` Latin-1**: ß and ÿ used to spill to the per-rune fallback; now handled inline with the same SIMD trick `fold_ice` already used for ß

## Pipeline ordering alignment

Used `fold_ice`'s branch order as the canonical layout and applied the same priority sequence to `fold_haswell`, `upper_haswell`, and `upper_ice`. After this all four functions share:

```
ASCII → Latin-1 → Cyrillic basic → Greek → Caseless 2-byte (Ice only)
     → Cyrillic Supplement → Universal 2-byte (Ice only)
     → 3-byte caseless → Nuskhuri/Asomtavruli → Mkhedruli/Mtavruli
     → Latin Extended Additional → Fullwidth → Serial drain
```

Also merged the four separate `upper_ice` Cyrillic Supplement branches (D2/D3/D1-ext/D4) into one combined branch — fixes a 17× regression on text with interleaved D2/D3/D4 leads.

## Bindings

- **Python** (`python/stringzilla.c`): added `utf8_case_upper` method on `Str` and module function. Buffer sized at `length * 6` with overflow check.
- **Rust** (`rust/stringzilla.rs`): added `utf8_case_upper` wrapper around extern `sz_utf8_case_upper`.

## Verification

| Test | Result |
|---|---|
| `serial vs CPython 3.14 str.upper()` on 1.1M codepoints | 0 diffs |
| `serial vs CPython 3.14 str.casefold()` on 1.1M codepoints | 0 diffs |
| `haswell vs serial` (upper + fold) on 1.1M codepoints | 0 diffs |
| `ice vs serial` (upper + fold) on 1.1M codepoints | 0 diffs |

Verified on Intel Xeon Platinum 8375C (Ice Lake-SP) for AVX-512 paths, AMD EPYC 7R13 (Zen 3) for AVX2 paths. Disassembly confirms `_haswell` functions emit no ZMM instructions; `_ice` functions emit no YMM. A single binary safely contains all backends, dispatched at runtime.

## Performance (Intel Xeon Platinum 8375C, 32 MiB, best of 7, GB/s)

| Workload | upper serial → haswell, ice | fold serial → haswell, ice |
|---|---|---|
| ASCII | 1.21 → 9.79 (×8.1), 9.08 (×7.5) | 1.22 → 9.47 (×7.8), 9.01 (×7.4) |
| Cyrillic | 0.29 → 3.78 (×13.1), 4.50 (×15.5) | 0.47 → 4.17 (×8.9), 3.90 (×8.3) |
| Greek | 0.43 → 2.82 (×6.6), 3.84 (×9.0) | 0.13 → 4.02 (×30.7), 3.85 (×29.3) |
| Mixed (Latin-1 accents) | 0.78 → 5.38 (×6.9), 6.24 (×8.0) | 0.91 → 6.43 (×7.0), 5.36 (×5.9) |
| Cyrillic Supplement (interleaved D2/D3/D4) | 0.11 → 3.11, 3.23 (×30 from serial) | 0.16 → 2.95 (×17.9), 2.81 (×17.0) |
| Mkhedruli | 0.53 → 1.85 (×3.5), 2.37 (×4.5) | 0.34 → 1.80 (×5.3), 1.90 (×5.6) |
| Mtavruli | 0.34 → 1.94 (×5.7), 2.77 (×8.1) | 0.62 → 1.68 (×2.7), 2.13 (×3.4) |
| CJK | 0.29 → 3.27 (×11.2), 2.45 (×8.4) | 0.32 → 3.29 (×10.4), 3.54 (×11.1) |

eshishki commented:
now it is better

eshishki commented:
understandable, i wanted to use stringzilla for levenshtein implementation, it was great
i decided to use lower utf8 implementation, it was great
i do not want to use icu for upper and stringzilla for lower, hence this pr

ashvardanian commented:
Gotcha, @rschu1ze! @eshishki, can you please reword the commits to match our naming convention and remove the new `scripts/*` files? I'll merge into a temporary feature branch and adjust with a new algorithm 🤗

eshishki commented:
@ashvardanian do you need need these test_case* files or nuke them?

ashvardanian commented:
@eshishki, ☢️!

eshishki committed:
Add: UTF-8 uppercase conversion with AVX-512

Commit:
Release: v4.6.0 [skip ci]
### Minor

- Add: Georgian fast path (e02cb00)

### Patch

- Fix: `inline static` warnings with C++23 modules (#287) (374adbf)
- Improve: Reduce startup overhead for `sz_find_byteset_haswell` (#293) (7f2899a)
- Fix: Missing Georgian dispatch (5f91e7e)
- Improve: Drop stack-protection in hashing on GCC (23801e4)
- Improve: Reduce repeated reviews (2e5784b)
- Improve: Faster `sz_size_bit_ceil` (4003057)
- Improve: Avoid ZMM-to-stack spill in Skylake comparisons (d94a010)
- Make: Use relative install path for C sources (690d775)

Algunenano committed:
Fix: OOB in `sz_find_skylake` / `sz_rfind_skylake` tail for null-byte needles (#312)

In the tail section of `sz_find_skylake` and `sz_rfind_skylake`, masked loads
(`_mm512_maskz_loadu_epi8`) zero out bytes beyond the valid haystack range.
The subsequent `_mm512_cmpeq_epi8_mask` comparisons are unmasked — they
compare all 64 lanes including the zeroed masked-off positions. When the
needle characters selected by `sz_locate_needle_anomalies_` are all `\0`
(e.g., needle = "\0\0\0\0"), the zeroed lanes falsely match, producing
spurious bits in `matches` at offsets beyond `h_length - n_length`.

For `n_length <= 3`, this returns an out-of-bounds pointer without any
validation. For `n_length > 3`, `sz_equal_skylake` reads past the haystack
boundary, causing a heap-buffer-overflow.

The fix is to AND `matches` with `mask` (the valid-position bitmask) before
entering the match loop, filtering out spurious matches from masked-off
positions. The same fix is applied to both `sz_find_skylake` and
`sz_rfind_skylake`.

Other implementations (serial, westmere, haswell, neon) are not affected
because they fall back to `sz_find_serial` for the tail instead of using
AVX-512 masked loads.

Reproducer:
    sz_find_skylake("AAAAAAAAAA", 10, "\0\0", 2)
    // Returns offset 9 (OOB) instead of NULL

Found via ClickHouse CI AST fuzzer (MSan build):
    SELECT count() FROM system.schema_inference_cache
    WHERE toNullable(65537) > countSubstrings(source, '\0\0\0\0')

Co-authored-by: Raúl Marín <664253+Algunenano@users.noreply.github.com>

belugabehr committed:
Fix: Signed char handling in `sz_find_byte_serial` (#308)

Closes #306

Co-Authored-By: David Mollitor <dmollitor@apache.org>
Co-Authored-By: David Mollitor <12578579+belugabehr@users.noreply.github.com>
Co-authored-by: Ash Vardanian <1983160+ashvardanian@users.noreply.github.com>

ashvardanian committed:
Fix: OOB read in `sz_order_skylake` tail for length-mismatched strings

The tail used masked loads with different masks for `a` and `b`, then an
unmasked `_mm512_cmpneq_epi8_mask`. When one string was shorter, its
masked-off lanes were zero, so the longer string's bytes in those lanes
compared as "not equal" against zero, producing spurious mismatch bits
beyond `min(a_length, b_length)`. The follow-up `a[first_diff]` /
`b[first_diff]` then read past the shorter buffer and could return a
wrong ordering whenever the random byte at the OOB index outranked the
real one — manifesting in `sz_sequence_argsort_with_insertion` as the
empty string being placed after non-empty strings starting with `\0`,
which made `test_sorting_algorithms` abort on the `"ab\0"` alphabet.

Restrict the compare to `a_mask & b_mask` (the bytes valid in both)
using the masked `_mm512_mask_cmpneq_epi8_mask`, mirroring how the head
section and `sz_equal_skylake` already handle masked tails.

ashvardanian committed:
Fix: Free `table_positions` in `sz_sequence_intersect_serial` and `_ice`

Both intersect implementations allocated the linear-probing hash table
via `alloc->allocate(...)` at function entry but never released it on
the success path, leaking `hash_table_slots * (sizeof(sz_size_t) +
sizeof(sz_u64_t))` bytes per call. LeakSanitizer flagged this as 6
direct leaks (~50 KiB total) in `test_intersecting_algorithms`. Add the
missing `alloc->free(...)` before the final `return sz_success_k;` in
each function. The `sz_bad_alloc_k` early return needs no free since
the allocation itself failed.

ashvardanian committed:
Make: Gitignore `.claude/` per-project settings

ashvardanian committed:
Chore: Clean-up dedicated `test_find_byte_serial_high_bytes` regression

The signed-char SWAR bug it covered (issue #306, PR #308) is structurally
prevented now that `sz_find_byte_serial` reinterprets its inputs as
`sz_u8_t const *` at function entry. The general find/rfind fuzz
already exercises the full byte range, so the dedicated function and
its top-level call are redundant.

Commit:
Release: v4.6.1 [skip ci]
### Patch

- Make: Gitignore `.claude/` per-project settings (37c2d8f)
- Fix: Free `table_positions` in `sz_sequence_intersect_serial` and `_ice` (087ad7f)
- Fix: OOB read in `sz_order_skylake` tail for length-mismatched strings (a3926e2)
- Fix: Signed char handling in `sz_find_byte_serial` (#308) (267c757)
- Fix: OOB in `sz_find_skylake` / `sz_rfind_skylake` tail for null-byte needles (#312) (7d72c96)

eshishki committed:
Fix: Drop bogus case folds for unassigned Unicode 16.0 codepoints

`sz_unicode_fold_codepoint_` mapped four codepoint groups that Unicode
16.0 marks as unassigned (Cn), producing folds out of thin air:
  - U+A7CE → U+A7CF
  - U+A7D2 → U+A7D3
  - U+A7D4 → U+A7D5
  - U+16EA0..U+16EB8 → U+16EBB..U+16ED3 (the entire "Beria Erfe" range)

Verified bit-perfect against `str.casefold` from CPython 3.14.2 /
unicodedata 16.0 over all 1.1M codepoints — 0 diffs remain.

eshishki committed:
Add: Haswell (AVX2) fast path for `sz_utf8_case_upper` and `_fold`

For text that is overwhelmingly ASCII (the common case in source code,
JSON, logs, English prose), the new Haswell backend uppercases / folds
32 bytes per iteration with `_mm256_min_epu8` and one `_mm256_sub_epi8`.
The first byte with the high bit set spills into the existing per-rune
serial mapper (`sz_unicode_upper_codepoint_` / `_fold_codepoint_`), so
multi-byte UTF-8 is handled correctly without porting all the AVX-512
script-specific transforms to AVX2.

The dynamic dispatcher picks Haswell when AVX-512 is unavailable —
covers AMD Zen 1-3 (Naples, Rome, Milan) and Intel Haswell..Cascade
Lake, where the previous build silently fell back to serial.

Verified bit-perfect against serial on AMD EPYC 7R13 (Zen 3) across all
1.1M Unicode codepoints — 0 diffs.

eshishki committed:
Ice upper: handle ß→SS and ÿ→Ÿ inside the Latin-1 SIMD path

`sz_utf8_case_fold_ice` already overwrites both bytes of `ß` (C3 9F)
with ASCII `'s'` directly in the SIMD path because the byte width
stays at 2. Mirror that for `_upper_ice`:

  * ß (C3 9F) → SS (53 53): `_mm512_mask_set1_epi8(..., 'S')` on both
    the lead and the second byte positions.
  * ÿ (C3 BF) → Ÿ (C5 B8): one masked add (+2 on the lead) + one
    masked sub (-7 on the second byte).

Before, both characters skipped the fast path entirely and drained
one rune at a time through the per-rune serial mapper. On text where
they appear regularly (German with `Straße`, French with `caf`/`naïve`,
etc.) this was a noticeable cliff. Benchmarked on Intel Xeon Platinum
8375C (Ice Lake-SP, 64 MiB input, best of 5):

    Mixed upper: 1.35 → 4.99 GB/s  (×3.7)

Still bit-perfect against serial across all 1.1M codepoints.

eshishki committed:
Add: AVX2 Latin-1 SIMD fast path + Haswell helpers

Lays the foundation for porting per-script transforms from the Ice
(AVX-512) path onto Haswell (AVX2). New helpers in the Haswell region:

  sz_haswell_eq_                — byte-wise cmpeq for a constant
  sz_haswell_first_invalid_     — AVX2 counterpart to sz_ice_first_invalid_
  sz_haswell_byte_mask_shl1_    — cross-lane byte-mask left-shift by 1
  sz_haswell_byte_mask_shr1_    — cross-lane byte-mask right-shift by 1
  sz_haswell_mask_sub_          — blend(v, v-delta, mask) wrapper
  sz_haswell_mask_add_          — blend(v, v+delta, mask) wrapper
  sz_haswell_mask_set1_         — blend(v, splat(c), mask) wrapper
  sz_haswell_store32_           — full 32-byte store (callers advance only by prefix length)
  sz_haswell_prefix_bytes_      — byte mask 0xFF in first N positions

First per-script branch: Latin-1 (C3). Mirrors what _ice does for upper
and fold, including the inline ß↔SS/ss trick (overwrite both bytes with
'S'/'s' since the byte width stays 2) and the ÿ→Ÿ irregular (C3 BF →
C5 B8 via lead+2, second-7) in upper.

Benchmarked on Intel Xeon Platinum 8375C (64 MiB, best of 5):

    Mixed   upper:  serial 0.86 → haswell 5.01 GB/s (x5.8), ice 5.29
    Mixed   fold :  serial 0.90 → haswell 5.99 GB/s (x6.7), ice 5.38

ASCII path unchanged at ~x6. Cyrillic still serial (next commit).
Verified bit-perfect haswell-vs-serial and ice-vs-serial across all
1.1M codepoints — 0 diffs.

eshishki committed:
Add: AVX2 Cyrillic basic (D0/D1) SIMD fast path for utf8 upper/fold

Mirrors the Ice (AVX-512) D0/D1 branch on AVX2. Covers Russian,
Ukrainian, Belarusian, Bulgarian, Serbian, Macedonian etc. — the
overwhelming majority of Cyrillic prose.

Upper transforms:
    D0 B0-BF (а-п) → D0 90-9F (А-П): second -0x20
    D1 80-8F (р-я) → D0 A0-AF (Р-Я): second +0x20, lead D1→D0 (-1)
    D1 90-9F (ѐ-џ) → D0 80-8F (Ѐ-Џ): second -0x10, lead D1→D0 (-1)

Fold transforms (inverse direction with lead +1 for the cross-D0/D1
pairs). D1 A0-BF (Cyrillic Supplement extended) is masked out of the
chunk so it can be handled by a future dedicated branch — keeping the
basic-Cyrillic path simple.

Benchmarked on Intel Xeon Platinum 8375C (64 MiB, best of 5):

    Cyrillic upper:  serial 0.27 → haswell 2.77 GB/s (x10.3), ice 3.69
    Cyrillic fold :  serial 0.41 → haswell 3.63 GB/s ( x8.9), ice 3.78

Verified bit-perfect haswell-vs-serial and ice-vs-serial across all
1.1M codepoints — 0 diffs.

eshishki committed:
Add: AVX2 Cyrillic Supplement (D1-ext/D2/D3/D4) for utf8 upper

Mirrors the four Ice (AVX-512) Cyrillic Supplement branches in a single
combined AVX2 branch. All transforms collapse to "second byte -1 (or
-0xF for U+04CF)" so the four lead-byte masks merge into one mask_sub.

Covers:
    D1 A0-BF odd       — U+0460-047F (old Slavonic etc.)
    D2 81              — irregular ҁ → Ҁ
    D2 8B-BF odd       — U+048A-04BF (Turkic, Cyrillic-using minorities)
    D3 82-8E even
    D3 8F              — irregular ӏ → Ӏ
    D3 91-BF odd       — U+04C0-04FF (Caucasian)
    D4 81-AF odd       — U+0501-052F (Komi, Mordvin, Chukchi)

Fold direction intentionally not added here — the Ice fold path also
doesn't cover the Cyrillic Supplement (it falls into the per-rune
serial drain), so we keep parity with Ice. Can be added later as a
joint Ice+Haswell improvement.

Verified bit-perfect against serial across 1.1M codepoints — 0 diffs.

eshishki committed:
Add: AVX2 Greek (CE/CF) SIMD fast path for utf8 upper/fold

Upper covers the full set of bicameral Greek transforms:
    CE B1-BF (α-ο)        → CE 91-9F (Α-Ο):     second -0x20
    CE AC    (ά)          → CE 86    (Ά):       second -0x26
    CE AD-AF (έ ή ί)      → CE 88-8A (Έ Ή Ί):   second -0x25
    CF 80-8B (π-ϋ)        → CE A0-AB (Π-Ϋ):     lead -1, second +0x20
    CF 82    (ς)          → CE A3    (Σ):       lead -1, second +0x21
    CF 8C    (ό)          → CE 8C    (Ό):       lead -1 only
    CF 8D-8E (ύ ώ)        → CE 8E-8F (Ύ Ώ):     lead -1, second +1

Fold mirrors Ice fold's narrower coverage (basic Α-Ω without tonos,
plus ς → σ final-sigma normalisation):
    CE 91-9F (Α-Ο)        → CE B1-BF: second +0x20
    CE A0-AB (Π-Ϋ)        → CF 80-8B: lead +1, second -0x20
    CF 82                 → CF 83:    second +1

Both branches skip-if-tricky when the chunk contains CE 90/B0
(SpecialCasing decompositions of ΐ/ΰ) or CF 90+ (Greek symbols /
archaic parity / irregular caps) — those defer to the per-rune
serial fallback.

Verified bit-perfect against serial across 1.1M codepoints — 0 diffs.

eshishki committed:
Add: AVX2 Mkhedruli, Nuskhuri, Latin Ext Additional, Fullwidth branches

Four more 3-byte SIMD branches in Haswell, mirroring upper_ice/fold_ice:

  Upper:
    - Georgian Mkhedruli (E1 83 90-BA, BD-BF) → Mtavruli (E1 B2 ...):
      second 0x83 → 0xB2 (+0x2F), skip chunks with E1 leads followed by
      non-0x83 to avoid corrupting Cherokee / Greek Ext / Latin Ext Add.
    - Georgian Nuskhuri (E2 B4 80-A5) → Asomtavruli (E1 82/83):
      composite lead/second/third transform, skip if E2 B4 third ≥ 0xA6
      (U+2D27, U+2D2D irregular).
    - Latin Extended Additional (E1 B8-BB) parity-based, skip E1 BA 96..9F.
    - Fullwidth EF BD 81-9A (ａ-ｚ) → EF BC A1-BA (Ａ-Ｚ).

  Fold:
    - Asomtavruli (E1 82/83) → Nuskhuri (E2 B4): inverse of Nuskhuri upper.
    - Latin Extended Additional (E1 B8-BB) parity-based: even third → +1.
    - Fullwidth EF BC A1-BA → EF BD 81-9A.

Each 3-byte branch uses the same chunk-boundary trim
(`last_lead_pos + 3 > length`) as the corresponding _ice branches.

Verified bit-perfect haswell-vs-serial across all 1.1M codepoints — 0 diffs.

eshishki committed:
Add: AVX2 3-byte caseless fast path (CJK/Hangul/Thai/...)

When the chunk contains no bicameral 3-byte leads (E1, E2, EA, EF) and
all non-ASCII bytes are valid 3-byte leads (E0, E3-E9, EB-EE) or
continuation bytes, the script is guaranteed caseless. Bulk-copy with
just ASCII a-z/A-Z transformation — no per-rune drain needed.

Covers Chinese (CJK Unified), Japanese (Hiragana/Katakana), Korean
(Hangul), Thai, Hindi/Devanagari, Arabic, Hebrew, and most other 3-byte
scripts without case mapping.

Same trim-hanging-lead pattern as the other 3-byte branches. Skipped
when E1/E2/EA/EF appear — those have case-bearing sub-ranges already
handled by the branches above and would otherwise corrupt them.

Verified bit-perfect haswell-vs-serial across all 1.1M codepoints — 0 diffs.

eshishki committed:
Add: AVX-512 and AVX2 Cyrillic Supplement + Mtavruli fold branches

Previously sz_utf8_case_fold_ice and _haswell had no SIMD branch for the
Cyrillic Supplement / Extended range (U+0460-052F) or Georgian Mtavruli
(U+1C90-1CBF). Cyrillic Suppl. went through Ice's universal codepoint-
unpack path; Mtavruli through the per-rune serial fallback. Both made
fold significantly slower than upper on the same workloads.

New branches (mirroring the existing upper branches):

  Cyrillic Supplement fold:
    D1 A0-BE even   → +1  (U+0460-047F)
    D2 80           → +1  (Ҁ → ҁ, irregular)
    D2 8A-BE even   → +1  (U+0480-04BF parity)
    D3 80           → +0xF (Ӏ → ӏ, irregular)
    D3 81-8D odd    → +1  (range 1: odd cap, even small)
    D3 90-BE even   → +1  (range 2: even cap, odd small)
    D4 80-AE even   → +1  (U+0500-052F)
  Excludes D1 80-9F (basic Cyrillic, handled earlier) and D4 B0+
  (Armenian, deferred to the universal path).

  Mtavruli → Mkhedruli fold (E1 B2 → E1 83):
    Single linear transform: second byte B2 → 83 (-0x2F). Skip chunks
    where any E1 B2 third byte is outside the Mtavruli range (U+1C80..
    U+1C8F Cyrillic Extended-C need irregular per-rune folds) or any E1
    lead is followed by non-B2 (Cherokee / Latin Ext Add / Greek Ext /
    basic Mkhedruli E1 83 in the same chunk).

Benchmarked on Intel Xeon Platinum 8375C (32 MiB, best of 5):
    Cyrillic-Suppl fold: serial 0.20 → haswell 2.82 (×14.1), ice 2.91 (×14.6)
    Mtavruli       fold: serial 0.54 → haswell 3.49 ( ×6.5), ice 2.38 ( ×4.4)

Verified bit-perfect haswell-vs-serial and ice-vs-serial across all
1.1M codepoints — 0 diffs.

eshishki committed:
Align fold_haswell, upper_haswell, upper_ice to fold_ice canonical ordering

Use fold_ice's branch ordering as the canonical pipeline (most mature
path from upstream). Apply the same priority sequence to all three
derivative functions for maintainability and consistency.

Canonical pipeline:
    ASCII → Latin-1 (C3) → Cyrillic basic (D0/D1) → Greek (CE/CF)
    → Caseless 2-byte D7-DF (Ice only) → Cyrillic Supplement (D1-ext/D2/D3/D4)
    → Universal 2-byte (Ice only) → 3-byte caseless
    → Nuskhuri/Asomtavruli (E2 B4 ↔ E1 82/83)
    → Mkhedruli/Mtavruli (E1 83 ↔ E1 B2)
    → Latin Extended Additional (E1 B8-BB) → Fullwidth (EF BC/BD)
    → Serial drain

Per-function moves:

  fold_haswell:
    - Swap Cyrillic Supplement ↔ Greek (Greek first).
    - Move 3-byte caseless from tail to after Cyrillic Supplement.

  upper_haswell:
    - Same swap and same 3-byte caseless move.
    - Swap Mkhedruli ↔ Nuskhuri so Nuskhuri runs first.

  upper_ice:
    - Move 3-byte caseless from position 2 (right after ASCII) to
      position 9 (after all 2-byte branches).
    - Move Greek from after Cyrillic Supplement to before it.
    - Swap Mkhedruli ↔ Nuskhuri so Nuskhuri runs first.

Verified bit-perfect haswell-vs-serial and ice-vs-serial across all
1.1M codepoints — 0 diffs.

Bench (Intel Xeon Platinum 8375C, 32 MiB, best of 7):
    ASCII    upper: serial 1.21 → haswell 9.79 (×8.1), ice 9.08 (×7.5)
    Cyrillic upper: serial 0.29 → haswell 3.78 (×13.1), ice 4.50 (×15.5)
    Greek    upper: serial 0.43 → haswell 2.82 (×6.6), ice 3.84 (×9.0)
    Mixed    upper: serial 0.78 → haswell 5.38 (×6.9), ice 6.24 (×8.0)
    CJK      upper: serial 0.29 → haswell 3.27 (×11.2), ice 2.45 (×8.4)

Note: heavily-interleaved scripts (a pattern alternating D2/D3/D4
Cyrillic Supplement leads byte-by-byte) make every chunk advance only
2-4 bytes through the per-lead fast paths, and the per-iteration
pre-checks dominate. This is a pipeline-architecture property, not a
regression — pure D2-only Cyrillic Supplement still runs at 3.5 GB/s
in Ice.

eshishki committed:
Ice upper: merge four Cyrillic Supplement branches into one

Replaces four separate Ice upper branches (D2, D3, D1-ext, D4) with a
single combined branch that detects all four lead bytes at once and
applies the unified "-1 (or -0xF) on the second byte" transform in one
mask_sub pass. Matches the structure already used in upper_haswell.

The old setup cut the chunk at the first non-matching lead byte each
time, so interleaved D2/D3/D4 text (the realistic shape of Caucasian
Cyrillic Supplement content) advanced 2-4 bytes per iteration and the
per-iteration pre-checks dominated. The combined branch processes the
whole chunk in one pass.

Benchmarked on Intel Xeon Platinum 8375C (32 MiB, best of 7):

    Mixed Cyrillic Suppl upper (D2/D3/D4 interleaved):
        before: serial 0.17 → ice 0.21 GB/s  (×1.3)
        after : serial 0.11 → ice 3.23 GB/s  (×29.4)

    Pure D2 (no interleaving):
        before: ice 3.50 GB/s
        after : ice 2.69 GB/s  (~25% regression — extra lead-mask ops
                                trade off against the huge mixed-pattern
                                win; pure single-lead Suppl content is
                                rare in real text)

Verified bit-perfect haswell-vs-serial and ice-vs-serial across all
1.1M codepoints — 0 diffs.

eshishki committed:
Upper: add Mtavruli passthrough fast path (Ice + Haswell)

Mtavruli (U+1C90-U+1CBF, E1 B2 90-BA / BD-BF) is already uppercase, so
upper(Mtavruli) = Mtavruli. Previously the upper pipeline had no fast
path for it: every codepoint fell through every script branch and into
the per-rune serial drain. Slower than serial because of the wasted
SIMD detection overhead.

Add a dedicated bulk-copy branch after Mkhedruli, in both Ice and
Haswell upper. Triggers only when every E1 lead is followed by B2 AND
every E1 B2 third byte is in the Mtavruli range — that excludes
Cyrillic Extended-C (E1 B2 80..8F, U+1C80-U+1C8F) which has irregular
lowercase folds and still needs the serial drain.

When it fires, the chunk is just an ASCII A-Z transform + 32/64-byte
store; the Mtavruli bytes pass through unchanged.

Benchmarked on Intel Xeon Platinum 8375C (32 MiB, best of 7):

    Mtavruli upper:
        before: serial 0.30 → haswell 0.23 (×0.8), ice 0.09 (×0.3)
        after : serial 0.34 → haswell 1.94 (×5.7), ice 2.77 (×8.1)

Verified bit-perfect haswell-vs-serial and ice-vs-serial across all
1.1M codepoints — 0 diffs.
